src/fsharp/TypeChecker.fs

```diff
@@ -15042,9 +15043,13 @@ module EstablishTypeDefinitionCores =
             let acc = 
                 match tycon.TypeAbbrev with 
                 | None -> acc
-                | Some ty -> 
-                    //if not cenv.isSig && not cenv.haveSig && (tycon.Accessibility <> taccessPublic || tycon.TypeReprAccessibility <> taccessPublic) then 
-                    //   errorR(Error(FSComp.SR.tcTypeAbbreviationMustBePublic(),tycon.Range))
+                | Some ty ->                    
```

Comment:
I think the right place for this check is ``CheckEntityDefn`` in ``PostInferenceChecks.fs`` using something like

    CheckTypeForAccess cenv env (fun () -> tycon.Name) tycon.Accessibility tycon.Range tycon.TypeAbbrev

This will check the type deeply, so ``type public X = List<MyPrivateType>`` will be detected correctly.

I'm really surprised we didn't have this check in, but there you go.  This will be a breaking change - so I guess we should consider making it a (strongly worded) warning.